homeassistant/components/modbus/sungrow_modbus_tcp_client.py

```diff
@@ -0,0 +1,84 @@
+"""Wrapping class around pymodbus ModbusTcpClient that adds support for Sungrow Solar Inverters with modbus AES encryption over TCP.
+
+Defaults to standard ModbusTcpClient if encryption is not supported by a sungrow device.
+"""
+from pymodbus.client.sync import ModbusTcpClient
+from Crypto.Cipher import AES
+
+PRIV_KEY = b"Grow#0*2Sun68CbE"
```

Comment:
Having a private key in a public file defies the purpose of a private key. Same goes with the crypto information. This should be handled in a private setup.

Reply:
Thinking about it, it could be a nice feature to add AES encryption generically to the tcpClient, and thereby allow other integrations to use it (I have a Huawei inverter that uses a mixture of encrypted and non-encrypted trafic.

Reply:
@janiversen, maybe we can keep an agnostic AES class within modbus and separate sungrow specific parts as you said. Can you give some pointers on how should I organize this (where to put things and how)? It is my first PR here.

Reply:
@janiversen this 'private key' is not user defined. It is a global/fixed key used by sungrow app. You can't change it.
Each inverter has a 'public key' that is XOR'ed with this 'public key' to generate the key itself.

Reply:
but we could retrieve it from '!secrets', no problem with that, just have to look up how to do this.

Reply:
I have put it on my plan to make a generic AES (I do not think we need a whole class) setup in modbus, but you are welcome to do  it.

Look at some of the other integrations that uses modbus (just search for modbus). It is very simple, but of course you need to add more files (you can get a scaffold made, and then adapt it). I am not at my computer right now so I cannot point you to an integration that uses modbus.

I know the private key is not user defined, but might very well change with a new version of sungrow, so correctly you cannot change it but sungrow can. I would really prefer to have keys in !secrets (and thus changeable) instead of in code.

Reply:
You also need to document these options, otherwise users will not know they exist (see documentation).

Test cases are important to ensure there are no sideeffect when e.g. modbus is upgraded. 

It seems you still have problems with flake8 and isort, all checks need to be successful.

Reply:
agreed. I'll put the key in !secrets and move the sungrow class outside modbus. Another time I'll try to factor out AES code and move it to modbus.

Reply:
Sound like a good plan. When you are ready for a new review, just mention me in a comment and I will take a look.